### PR TITLE
[#996] updated min versions with test function to ensure our regex is…

### PR DESCRIPTION
**Pull request by joehiggi1758**
… caputring all edge cases

# Pull Request Checklist

Below is a simple checklist but please do not hesitate to ask for assistance!

- [x] Fork, clone, and checkout the newest version of the code
- [x] Create a new branch
- [x] Make necessary code changes
- [x] Install `black` (i.e., `python -m pip install black` or `conda install -c conda-forge black`)
- [x] Install `flake8` (i.e., `python -m pip install flake8` or `conda install -c conda-forge flake8`)
- [x] Install `pytest-cov` (i.e., `python -m pip install pytest-cov` or `conda install -c conda-forge pytest-cov`)
- [x] Run `black .` in the root stumpy directory
- [x] Run `flake8 .` in the root stumpy directory
- [x] Run `./setup.sh && ./test.sh` in the root stumpy directory
- [x] Reference a Github issue (and create one if one doesn't already exist)

### CODE REVIEW

File: min_versions.py [hunk 1]
```diff
@@ -143,6 +143,23 @@ def get_min_scipy_version(min_python, min_numpy):
     return df.SciPy_version
 
 
+def match_str_version(line, pkg_name):
```
Comment (seanlaw) [inline min_versions.py:146]:
```suggestion
def match_pkg_version(line, pkg_name):
```

File: min_versions.py [hunk 2]
```diff
@@ -153,15 +170,7 @@ def find_pkg_mismatches(pkg_name, pkg_version, fnames):
...
-                matches = re.search(
-                    rf"""
-                        {pkg_name}  # Package name
-                        [=><:"\'\[\]]+  # Zero or more special characters
-                        (\d+\.\d+[\.0-9]*)  # Capture "version" in `matches`
-                        """,
-                    l,
-                    re.VERBOSE,  # Ignores all whitespace in pattern
-                )
+                matches = match_str_version(l, pkg_name)
```
Comment (seanlaw) [inline min_versions.py:173]:
```suggestion
                matches = match_pkg_version(l, pkg_name)
```

File: min_versions.py [hunk 3]
```diff
@@ -170,6 +179,39 @@ def find_pkg_mismatches(pkg_name, pkg_version, fnames):
...
+
+    lines = [
+        "Programming Language :: Python :: 3.8",
+        "STUMPY supports Python 3.8",
+        "python-version: ['3.8']",
+        'requires-python = ">=3.8"',
+        "numba>=0.55.2",
+    ]
+
+    matches_found = []
```
Comment (seanlaw) [inline min_versions.py:202]:
```suggestion
```

File: min_versions.py [hunk 4]
```diff
@@ -170,6 +179,39 @@ def find_pkg_mismatches(pkg_name, pkg_version, fnames):
...
+    for pkg_name, pkg_version in pkgs.items():
+
```
Comment (seanlaw) [inline min_versions.py:204]:
```suggestion
```

File: min_versions.py [hunk 5]
```diff
@@ -170,6 +179,39 @@ def find_pkg_mismatches(pkg_name, pkg_version, fnames):
     return pkg_mismatches
 
 
+def test_pkg_mismatch_regex():
+    """
+    Validation function for the package mismatch regex
+    """
+    pkgs = {
```
Comment (seanlaw) [inline min_versions.py:186]: I guess not all of these packages actually trigger a failure?

File: min_versions.py [hunk 6]
```diff
@@ -170,6 +179,39 @@ def find_pkg_mismatches(pkg_name, pkg_version, fnames):
...
+    for pkg_name, pkg_version in pkgs.items():
+
+        for line in lines:
+            matches = match_str_version(line, pkg_name)
+
+            if matches:
+                matches_found.append(matches.group(0))
+
+    if len(matches_found) != len(lines):
+        raise ValueError("Edge case not matched by regex")
```
Comment (seanlaw) [inline min_versions.py:212]: I don't think that this is a good inequality check as it assumes that each of the matches found correspond to each line found in lines. If an error is made in our code/regex then it might be possible, for example, that a single line could produce multiple matches. 

Perhaps it would be "better" if we swapped the order of the `for` loops so that we iterate over lines first and then packages and `break` if a match is found but raise an error if no match is found and then provide a specific error message that identifies precisely which `line` is not being discovered by the regex.

Thoughts?

**Review by seanlaw — CHANGES_REQUESTED** [review]
@joehiggi1758 I've left some comments for you to consider

**Comment by seanlaw** [comment]
@joehiggi1758 I also recommend making your PR titles:

1. More brief/less verbose
2. Or preferably `Fixed #XYZ Name of Corresponding Issue`

The `Fixed #XYZ` corresponds to issue number `XYZ` and so when this PR gets closed/merged, it will also trigger the closing of issue `XYZ` automatically.

**Review by joehiggi1758 — COMMENTED** [review]
Applied requested changes and fixed/simplified `test_pkg_mismatch_regex()`! @seanlaw

File: min_versions.py (hunk 3 above)
Comment (seanlaw) [reply min_versions.py]: I think this was simply to remove/delete the line

File: min_versions.py (hunk 4 above)
Comment (seanlaw) [reply min_versions.py]: Please delete a blank line

File: min_versions.py [hunk 7]
```diff
@@ -170,6 +179,39 @@ def find_pkg_mismatches(pkg_name, pkg_version, fnames):
...
+        match_found = False
+        for pkg_name, pkg_version in pkgs.items():
+            matches = match_pkg_version(line, pkg_name)
+
+            if matches:
+                match_found = True
+                break
+
+        if not match_found:
+            raise ValueError(f"Edge case {line} not matched by regex")
```
Comment (seanlaw) [inline min_versions.py:212]: Maybe change this error message to `f'Package mismatch regex fails to cover/match "{line}"'`

**Comment by seanlaw** [comment]
Alright @joehiggi1758 everything looks good to me! Shall I merge this PR?

**Comment by joehiggi1758** [comment]
> Alright @joehiggi1758 everything looks good to me! Shall I merge this PR?

Sounds great, that works for me!

**Comment by seanlaw** [comment]
@joehiggi1758 Thank you for this PR!

**Comment by joehiggi1758** [comment]
@seanlaw thank you for the opportunity!
